### ISSUE TITLE
Remote control: Support link feature

### ISSUE THREAD

**Issue by Maxou44** [ncw/rclone#3042]
Hello here!

Thanks for this amazing software ✌

#### What problem are you are trying to solve?

I would like to get the download link from the remote control API, actually it's not possible, this feature is a remote control implementation of the rclone link command.

https://rclone.org/commands/rclone_link/

#### How do you think rclone should be changed to solve that?

I think Rclone could add a new endpoint to support it, we can imagine something like that:

```
{
   path: 'gdrive:/myfile.mp4'
   url: 'https://drive.google.com/.../myfile.mp4'
   encrypted_content: false,
   encrypted_title: false
}
```

It could also be nice if Rclone could generate links even if the file/title is encrypted, we can imagine 3rd party software to decrypt it.

**Comment by ncw** [comment]
Try this which is a straight forward implementation of `rclone link`

https://beta.rclone.org/branch/v1.46.0-068-gb743d689-fix-3042-rc-publiclink-beta/ (uploaded in 15-30 mins)

Here is the help

### operations/publiclink: Create or retrieve a public link to the given file or folder.

This takes the following parameters

- fs - a remote name string eg "drive:"
- remote - a path within that remote eg "dir"

Returns

- url - URL of the resource

See the [link command](https://rclone.org/commands/rclone_link/) command for more information on the above.

Authentication is required for this call.

> It could also be nice if Rclone could generate links even if the file/title is encrypted, we can imagine 3rd party software to decrypt it.

I didn't understand what you mean here? Do you mean crypt doesn't support public link?

**Comment by Maxou44** [comment]
Exactly, I'm wrong?
Is it possible to add informations in the output to say if the file is encrypted ?

**Comment by ncw** [comment]
Is this the problem?

```
$ rclone link TestCryptDrive:hello.txt
2019/03/13 14:50:24 Failed to link: Encrypted drive 'TestCryptDrive:hello.txt' doesn't support public links
```

**Comment by Maxou44** [comment]
Yes it's this error, do you think it could be possible to show the public link of the encrypted data with informations in the JSON output?

```
{
   url: 'https://drive.google.com/.../myfile.mp4'
   encrypted_content: true,
   encrypted_title: true
}
```

**Comment by ncw** [comment]
> Yes it's this error, do you think it could be possible to show the public link of the encrypted data with informations in the JSON output?
> 
> ```
> {
>    url: 'https://drive.google.com/.../myfile.mp4'
>    encrypted_content: true,
>    encrypted_title: true
> }
> ```

You can use [operations/list](https://rclone.org/rc/#operations-list-list-the-given-remote-and-path-in-json-format) to discover encrypted names for unencrypted names and to see if it is encrypted or not.

Did you try the beta above with `operations/link`?

https://beta.rclone.org/branch/v1.46.0-068-gb743d689-fix-3042-rc-publiclink-beta/

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.47

**Comment by ncw** [comment]
Here is a beta with link through crypt implemented - please give it a go!

https://beta.rclone.org/branch/v1.46.0-112-g716bb755-fix-2895-optional-methods-beta/ (uploaded in 15-30 mins)

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.48
